### ISSUE TITLE
Add config variable to print integers as hex

### ISSUE THREAD

**Issue by mtijanic** [bpftrace/bpftrace#3293]
### Is your feature request related to a problem? Please describe.

In some domains, decimal integers are hardly ever used and most values only make sense when examined as hex/octal/binary. Think bit flags, special magic numbers like 0xdeadbeef, etc. bpftrace always prints ints out as decimal.

A very common action (for me at least) is to print out the fields of some structure. #2557 is amazing in that regard, but because it prints out all integer-y arguments as decimals I ended up not using it anywhere.

### Describe the solution you'd like

Add a new config variable that controls the default formatting of integer types, globally, everywhere.

### Describe alternative solutions or features you've considered

I've settled on running bpftrace scripts through `cpp(1)` first, and using an X-macro like pattern to define structures, which allows me to generate code that "visits" every field and print it out with `printf("%x")`. Resulting code has additional moving parts (cpp) and ends up being about 2x larger than it should be.

**Comment by viktormalik** [comment]
This looks like a useful feature. I think that it would be limited to the `print` builtinl, right?

Anyways, it should be quite easy to add a new config option like that.

**Comment by mtijanic** [comment]
> I think that it would be limited to the print builtinl, right?

Yes. There's two places where I'd really love to see this:
1. `print(*$some_struct_ptr);`
2. `@[$some_int] = $some_other_int; print(@);`

IMO (1) is 90% of the issue, and the $some_int part of (2) is another 9%. For $some_other_int, I'm guessing consistency is the best choice so same rules as well?

For everything else, you can use `printf()` and specify the format manually.

Another alternative might be to pass an additional argument to `print()` to choose how to format ints for that call only?

**Comment by jordalgo** [comment]
> Another alternative might be to pass an additional argument to print() to choose how to format ints for that call only?

I had this thought as well but it map printing already has additional arguments [`top` and `div`](https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#print). That's not to say we can't add another one but maybe it's just easier to have a config var.

**Comment by ajor** [comment]
It would be nice to be able to control this individually for each `print()` call, but I do see the issue with too many arguments.

Just throwing out an idea that I haven't really thought through, but what if we introduced named arguments, e.g.:
```
print(@mymap, top=5);
print(@mymap, fmt=hex);
```

(maybe we could also use it for async map printing: #3028)

**Comment by ajor** [comment]
And maybe the up-front map declaration syntax could also somehow specify how to print the keys/values: #2954.

**Comment by jordalgo** [comment]
> what if we introduced named arguments

Yes, love this idea. I think we should try to generally move in this direction. Positional arguments often end up being just a pain for backward compatibility and are just less easy to grok.

That being said I think a config might still be good for this particular use case to reduce repetition if there are multiple print statements.

> And maybe the up-front map declaration syntax could also somehow specify how to print the keys/values: https://github.com/bpftrace/bpftrace/issues/2954.

IDK. I'm less keen on this idea just because the printing of maps seems separate from defining the map itself. It also just seems like a strange middle ground between a global config and a local `print` that specifies the print style in an named arg.

**Comment by ajor** [comment]
Also related: #2952, which describes a way in which maps can be printed in hexadecimal now.

**Comment by mtijanic** [comment]
Thinking about it some more, IMO the best default behavior would be to print `unsigned` types as hex, while `signed` types get printed as decimal. That would be too much of a breaking change for actual defaults, but a config variable that turns on that behavior gets you the best of all worlds.

**Comment by viktormalik** [comment]
Just a couple of thoughts on what was said in this thread:

> what if we introduced named arguments

I like the idea. Also, it would make it easier to allow optional parameters to user-defined functions, if we want to add them one day.

> > And maybe the up-front map declaration syntax could also somehow specify how to print the keys/values: https://github.com/bpftrace/bpftrace/issues/2954.

> IDK. I'm less keen on this idea just because the printing of maps seems separate from defining the map itself.

Agreed with @jordalgo here. The printing of maps should be independent from the map itself.

> Thinking about it some more, IMO the best default behavior would be to print `unsigned` types as hex, while `signed` types get printed as decimal. That would be too much of a breaking change for actual defaults, but a config variable that turns on that behavior gets you the best of all worlds.

I would avoid having a different default behaviour for signed and unsigned types.

Overall, I'd go with the config option for now. I'd suggest something like `print_int_format` with possible values either from an "enum" (`"dec"`, `"hex"`, `"oct"`, ...) or with a `printf`-like specifier. Then, we could add it as an optional `print` param once we introduce calls with named arguments.

**Comment by viktormalik** [comment]
> Bear in mind that there are already ways to print integers as hex in the given use-cases, they just take a bit more manual work with `printf` (a function designed for formatted printing).

Yeah, but when you want to print a huge struct (`task_struct` for instance), you really don't want to give the full `printf` specifier. In addition, it won't be portable between kernel versions.

**Comment by JacobLef** [comment]
Is this still open to be worked on? I am looking to get into contributing here and wanted to pickup a small task with my limited time currently.

**Comment by jordalgo** [comment]
@JacobLef Thanks for your interest but I think I'm going to close this issue - after 2 years I finally came around and agree with @ajor that we should not be adding new config variables whenever possible.

**Comment by JacobLef** [comment]
@jordalgo are there any good issues that I can get started on in the issues list? was trying to find a relaively simple one to be my first.

**Comment by jordalgo** [comment]
Here is a possible candidate:
https://github.com/bpftrace/bpftrace/issues/4185

Truth be told, I need to spend time updating the issues - I'm a bit backlogged at the moment
